ext/openssl/openssl.c

```diff
@@ -1355,32 +1430,73 @@ static X509 * php_openssl_x509_from_zval(zval * val, int makeresource, zend_reso
 		return NULL;
 	}
 
-	if (makeresource && resourceval) {
-		*resourceval = zend_register_resource(cert, le_x509);
-	}
 	return cert;
 }
 
+/* {{{ php_openssl_x509_from_param
+	Given a parameter, extract it into an X509 object.
+	The parameter can be:
+		. X509 object created using openssl_read_x509()
+		. if it starts with file:// then it will be interpreted as the path to that cert
+		. it will be interpreted as the cert data
+*/
+static X509 *php_openssl_x509_from_param(zend_object *cert_obj, zend_string *cert_str) {
+	if (cert_obj) {
+		return x509_certificate_from_obj(cert_obj)->x509;
+	}
+
+	ZEND_ASSERT(cert_str);
+
+	return php_openssl_x509_from_str(cert_str);
+}
+/* }}} */
+
+static X509 *php_openssl_x509_from_zval(zval *val, int *is_cert_object)
+{
+	*is_cert_object = 0;
+
+	if (Z_TYPE_P(val) == IS_OBJECT && Z_OBJCE_P(val) == x509_certificate_ce) {
+		*is_cert_object = 1;
+
+		return x509_certificate_from_obj(Z_OBJ_P(val))->x509;
+	}
+
+	if (Z_TYPE_P(val) != IS_STRING && Z_TYPE_P(val) != IS_OBJECT) {
+		return NULL;
+	}
+
+	if (!try_convert_to_string(val)) {
+		return NULL;
+	}
+
+	return php_openssl_x509_from_str(Z_STR_P(val));
+}
 /* }}} */
 
 /* {{{ Exports a CERT to file or a var */
 PHP_FUNCTION(openssl_x509_export_to_file)
 {
-	X509 * cert;
-	zval * zcert;
+	X509 *cert;
+	zend_object *cert_obj;
+	zend_string *cert_str;
+
 	zend_bool notext = 1;
 	BIO * bio_out;
 	char * filename;
 	size_t filename_len;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "zp|b", &zcert, &filename, &filename_len, &notext) == FAILURE) {
-		RETURN_THROWS();
-	}
+	ZEND_PARSE_PARAMETERS_START(2, 3)
+		Z_PARAM_STR_OR_OBJ_OF_CLASS(cert_str, cert_obj, x509_certificate_ce)
```

Comment:
so this actually prevents stringified object to be passed which has been supported so this breaking the functionality somehow.

Reply:
I think this should be kept as zval

Reply:
This still allows stringified objects per PHP's usual type declaration semantics. (In weak typing mode only of course, but that's a language consistency requirement.)

Reply:
Ah yeah that's good then. I think we can drop that `object OR string` condition in that case.